src/future/server/shutdown.rs

```diff
@@ -0,0 +1,184 @@
+            futures::Either::A(futures::ok(()))
+        } else {
+            futures::Either::B(AlwaysOkUnit(rx))
+        };
+        ShutdownFuture { inner: inner }
+    }
+}
+
+#[derive(Debug)]
+pub struct Watcher {
+    shutdown_rx: stream::Take<mpsc::UnboundedReceiver<oneshot::Sender<()>>>,
+    connections: unsync::mpsc::UnboundedReceiver<connection::Action>,
+    queued_error: Option<()>,
+    shutdown: Option<oneshot::Sender<()>>,
+    done: bool,
+    num_connections: u64,
+}
+
+impl Watcher {
+    pub fn triple() -> (connection::Tracker, Shutdown, Self) {
+        let (connection_tx, connections) = connection::Tracker::pair();
+        let (shutdown_tx, shutdown_rx) = mpsc::unbounded();
+        (connection_tx,
+         Shutdown { tx: shutdown_tx },
+         Watcher {
+            shutdown_rx: shutdown_rx.take(1),
+            connections: connections,
+            queued_error: None,
+            shutdown: None,
+            done: false,
+            num_connections: 0,
+        })
+    }
+
+    fn process_connection(&mut self, action: connection::Action) {
+        match action {
+            connection::Action::Increment => self.num_connections += 1,
+            connection::Action::Decrement => self.num_connections -= 1,
+        }
+    }
+
+    fn poll_shutdown_requests(&mut self) -> Poll<Option<()>, ()> {
+        Ok(Async::Ready(match try_ready!(self.shutdown_rx.poll()) {
+            Some(tx) => {
+                debug!("Received shutdown request.");
+                self.shutdown = Some(tx);
+                Some(())
+            }
+            None => None,
+        }))
+    }
+
+    fn poll_connections(&mut self) -> Poll<Option<()>, ()> {
+        Ok(Async::Ready(match try_ready!(self.connections.poll()) {
+            Some(action) => {
+                self.process_connection(action);
+                Some(())
+            }
+            None => None,
+        }))
+    }
+
+    fn poll_shutdown_requests_and_connections(&mut self) -> Poll<Option<()>, ()> {
+        if let Some(e) = self.queued_error.take() {
+            return Err(e)
+        }
+
+        match try!(self.poll_shutdown_requests()) {
+            Async::NotReady => {
+                match try_ready!(self.poll_connections()) {
+                    Some(()) => Ok(Async::Ready(Some(()))),
+                    None => Ok(Async::NotReady),
+                }
+            }
+            Async::Ready(None) => {
+                match try_ready!(self.poll_connections()) {
+                    Some(()) => Ok(Async::Ready(Some(()))),
+                    None => Ok(Async::Ready(None)),
+                }
+            }
+            Async::Ready(Some(())) => {
+                match self.poll_connections() {
+                    Err(e) => {
+                        self.queued_error = Some(e);
+                        Ok(Async::Ready(Some(())))
+                    }
+                    Ok(Async::NotReady) | Ok(Async::Ready(None)) => {
```

Comment:
Relatedly, there was an RFC to allow writing `Ok(Async::Ready(None | Some(())))` but I think it was postponed...